### DISCUSSION TITLE
Inserting default values when using subqueries

### DISCUSSION THREAD

**Discussion by magthe** [morphismtech/squeal#359] [Q&A] [answered]
How do I get the columns' default values when I'm inserting based on a sub-query?

In the schema I have I'd use this SQL to create a row each in `processes` and `products` where the latter refers to the former:

```sql
with proc as
    (insert into processes (type)
     values ('farm')
     returning id)
insert into products (proc_id, eid, name)
select id, ($1 :: uuid), ($2 :: text)
  from proc
```

trying to translate that into squeal I've got this so far

```haskell
sql = with
        (insertInto #processes
            (Values_ (Default `as` #id :*
                      Default `as` #eid :*
                      Default `as` #generation :*
                      Set "farm" `as` #type :*
                      Default `as` #extras))
            OnConflictDoRaise
            (Returning_ #id) `as` #proc)
    (insertInto_ #products
        (Subquery (select_ (undefined `as` #id :*
                            (param @1) `as` #eid :*
                            (param @2) `as` #name :*
                            #proc ! #id `as` #proc_id :*
                            undefined `as` #alloc)
                    (from (common #proc)))))
```

AFAIU there's no way getting around giving values for all columns (right???), and that's easy for the `insert` in the `with`-clause, but what do I use instead of `undefined` in the lower `insert`?

**Comment by echatav** [comment, marked as answer]
Instead of `Subquery` use `Select`. That will let you use `Default` in your subquery.

https://hackage.haskell.org/package/squeal-postgresql-0.9.1.3/docs/Squeal-PostgreSQL-Manipulation-Insert.html#t:QueryClause

Here's an example:

```Haskell
>>> :{
let
  insertUser :: Statement DB User ()
  insertUser = manipulation $ with (u `as` #u) e
    where
      u = insertInto #users
        (Values_ (Default `as` #id :* Set (param @1) `as` #name))
        OnConflictDoRaise (Returning_ (#id :* param @2 `as` #email))
      e = insertInto_ #emails $ Select
        (Default `as` #id :* Set (#u ! #id) `as` #user_id :* Set (#u ! #email) `as` #email)
        (from (common #u))
:}
```

**Reply by magthe** [reply]
The organisation of the example made me wonder, is it possible to lift out the definition of `u` so that it could be re-used by other statements?

So the usage would look something like this

```haskell
insertUser :: Statement DB User ()
insertUser = manipulation $ with ((u <passing 2 params here>) `as` #u) e
  where
    e = insertInto_ #emails $ Select
      (Default `as` #id :* Set (#u ! #id) `as` #user_id :* Set (#u ! #email) `as` #email)
      (from (common #u))
```

**Reply by echatav** [reply]
Of course, being able to reuse queries and other bits compositionally is the point.

**Reply by magthe** [reply]
Yes... but how? I don't think I've come across any examples of it.

**Reply by echatav** [reply]
First, you don't need to explicitly pass the parameters to `u` since parameters are passed out-of-line, later, with `executeParams`. You can just take the definition of `u` out of that (Haskell) where clause and make it a top level `Manipulation`. Then `u` can be reused in the definition of `insertUser` or anywhere else you can use a `Manipulation` with`u`'s type.

**Reply by echatav** [reply]
Composition of out-of-line parameters works differently than composition of functions. All the queries and manipulations and expressions that get composed together must have the same `params :: [NullType]`. But, if one piece of sql doesn't have any `$n`s in it then it can be polymorphic in `params` and hence compose with _any_ other piece of sql which _does_ have `$n`s and cannot be fully polymorphic in `params`. For the `u` manipulation, `params ~ 'NotNull 'PGtext ': 'NotNull 'PGtext ': params0` or something like that so `u` isn't fully polymorphic in `params`. Anything `u` gets composed into will have the same restriction on its `params`.

**Reply by echatav** [reply]
Generally, the more polymorphic the sql type, the fewer sql variables/aliases it uses, the more composable it will be. That is why an `Expr` can be used _anywhere_ that you need an `Expression`. It is simply an `Expression` which is fully polymorphic/universally quantified/closed in _all_ of its variable namespace type parameters.

```Haskell
type Expr x = forall grp lat with db params from. Expression grp lat with db params from x
```

**Reply by magthe** [reply]
I'm not sure I follow all of that, but if `u` uses `param @1` and expects it to be a `PGtext`, how do I compose that with an expression that uses 3 parameters, and it's the 2nd, `param @2`, that should be passed to `u`?

Are you saying that parameter positions need to match up exactly between `u` and users of `u`?

**Reply by echatav** [reply]
When you use a Squeal param you set it to a particular value of `n`, e.g. `param @n` so you cannot reassign it to a different value. The underlying `ByteString` would be different then. Here's the `insertUser` example but with `u` and `e` lifted to top level.

```Haskell
insertUser :: Manipulation '[] DB '[ 'NotNull 'PGtext, 'Null 'PGtext] '[]
insertUser = with (u `as` #u) e

u :: Manipulation '[] DB '[ 'NotNull 'PGtext, 'Null 'PGtext] '["id" ::: 'NotNull 'PGint4, "email" ::: 'Null 'PGtext]
u = insertInto #users
  (Values_ (Default `as` #id :* Set (param @1) `as` #name))
  OnConflictDoRaise (Returning_ (#id :* param @2 `as` #email))

e :: Manipulation '["u" ::: '["id" ::: 'NotNull 'PGint4, "email" ::: 'Null 'PGtext]] DB params '[]
e = insertInto_ #emails $ Select
  (Default `as` #id :* Set (#u ! #id) `as` #user_id :* Set (#u ! #email) `as` #email)
  (from (common #u))
```

Since `e` doesn't refer to any parameters it can be fully polymorphic in `params`. But `u` and `insertUser` (because it has `u` in it) has set `param @1` and `param @2` with specific types. We typed their `params` lists as lists of length `2` but they could have been longer like `('NotNull 'PGtext ': 'Null 'PGtext ': moreParams)`. Since we have not yet set any higher parameters, it can be polymorphic in `moreParams`.

> Are you saying that parameter positions need to match up exactly between u and users of u?

Yes, I think that's right. If you want to interpolate a string with a `$1 :: text` inside of a SQL query then that query had better either not have a `$1` in it already or if it does it better also be typed as `text`.

**Comment by magthe** [comment]
Oh, that's very nice. The generated SQL is surprisingly short too as all the `Default`s disappear from the `SELECT`!

**Reply by echatav** [reply]
If only I was clever enough to figure out how to make `Default` disappear from user code!

I used some ideas from Squeal's `Default` in [this new library](https://hackage.haskell.org/package/singletons-default). The idea for how to make default implicit (empty syntax) eludes me though.

**Reply by echatav** [reply]
Come to think of it, Squeal has the `Updatable` constraint which lets you leave out some columns (and update them out of order) but with updates you don't have to update any fields, default or otherwise. It's trickier I think to check that your insert fields are _all_ the `NoDef` fields and _some_ of the `Def` fields.
